Source/DafnyCore/Verifier/BoogieGenerator.Methods.cs

```diff
@@ -1474,9 +1477,12 @@ public partial class BoogieGenerator {
       var callee = new List<Expr>();
       var caller = new List<Expr>();
 
+      FunctionCallSubstituter sub = null;
+
       for (int i = 0; i < N; i++) {
         Expression e0 = calleeDecreases[i];
-        Expression e1 = Substitute(contextDecreases[i], null, substMap, typeMap);
+        sub ??= new FunctionCallSubstituter(substMap, typeMap, T, I);
```

Comment:
You can leave this as is, but I don't think the extra lines here are worth the saving of creating `sub` only once and only if needed. If you agree, you could either always initialize `sub` before the loop (which is more costly only when `N` is `0`) or create a new `sub` in each iteration (which is more costly only if `N` is `2` or more). I'd probably go for the latter.

Reply:
I tried inlining the construction of the substituter, but even I think it a little much for a single line. Without it, we're only saving a single line.